File: solutions/java/secret-handshake/1/src/main/java/HandshakeCalculator.java
import java.util.Arrays;
import java.util.List;

class HandshakeCalculator {

    private static Signal[] SIGNALS = Signal.values();

    List<Signal> calculateHandshake(int number) {
        Signal[] signals = new Signal[5];
        int index = 0;
        for (int i = 0; i < SIGNALS.length; ++i) {
            if (((number >> i) & 1) == 1) {
                signals[index++] = SIGNALS[i];
            }
        }
        if (((number >> SIGNALS.length) & 1) == 1) {
            for (int start = 0, end = index - 1; start < end; ++start, --end) {
                Signal temp = signals[end];
                signals[end] = signals[start];
                signals[start] = temp;
            }
        }
        return Arrays.asList(Arrays.copyOf(signals, index));
    }

}

File: solutions/kotlin/secret-handshake/1/src/main/kotlin/HandshakeCalculator.kt
object HandshakeCalculator {
    private val SIGNALS: Array<Signal> = enumValues<Signal>()
    
    fun calculateHandshake(number: Int): List<Signal> = SIGNALS
        .filterIndexed { index, _ -> number and (1 shl index) > 0 }
        .let { if (number and (1 shl SIGNALS.size) > 0) it.reversed() else it }
}
